lib/sequence.js

```diff
@@ -39,26 +38,41 @@ Sequence.prototype._run = function (fromIndex) {
 
 		var result = this.tests[i].run();
 
+		// if a Promise returned, we don't need to check for Promises after this test
+		// so we can just use Promise.each() on the rest of the tests
```

Comment:
Similar issue in <https://github.com/sindresorhus/ava/pull/542#issuecomment-189892779>.

At the risk of splitting up the discussion too much, one thing I've done in the past is to return a wrapper result from mixed synchronicity functions, either with a `promise` property or a `value` property. Then you can do `if (result.promise)` without having to do further type checking.